### DISCUSSION TITLE
Foxglove Studio does not appear to understand json-schema $ref in schemas

### DISCUSSION THREAD

**Discussion by kenaniah** [foxglove/community#971] [Bugs]
**Description**

- Version: 1.86.1 (darwin)
- Platform: MacOS 
- Data source (e.g. bag file, mcap file, rosbridge, ROS 1/2 native): MCAP file

It appears as though Foxglove Studio does not properly parse json-schema schemas. The following reproduction steps suggest that `$ref` is not handled, and the Foxglove Studio documentation does not make it clear which features of json-schema are supported or not supported.

**Steps To Reproduce**

1. Add a channel with a json-schema that uses references, for example:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "TracksMessage",
  "type": "object",
  "required": [
    "objects",
    "total"
  ],
  "properties": {
    "objects": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/TrackObject"
      }
    },
    "total": {
      "type": "integer",
      "format": "uint",
      "minimum": 0.0
    }
  },
  "definitions": {
    "TrackObject": {
      "type": "object",
      "required": [
        "altitude",
        "first_seen",
        "heading",
        "id",
        "last_seen",
        "latitude",
        "longitude",
        "source_type"
      ],
      "properties": {
        "altitude": {
          "type": "number",
          "format": "double"
        },
        "first_seen": {
          "type": "string",
          "format": "date-time"
        },
        "heading": {
          "type": "number",
          "format": "double"
        },
        "id": {
          "type": "string"
        },
        "last_seen": {
          "type": "string",
          "format": "date-time"
        },
        "latitude": {
          "type": "number",
          "format": "double"
        },
        "longitude": {
          "type": "number",
          "format": "double"
        },
        "on_ground": {
          "type": [
            "boolean",
            "null"
          ]
        },
        "source_type": {
          "$ref": "#/definitions/SourceType"
        },
        "tail_number": {
          "type": [
            "string",
            "null"
          ]
        }
      }
    },
    "SourceType": {
      "description": "Source type",
      "oneOf": [
        {
          "description": "ADS-B",
          "type": "string",
          "enum": [
            "ADSB"
          ]
        },
        {
          "description": "FAA Terrestrial",
          "type": "string",
          "enum": [
            "ASDI"
          ]
        }
      ]
    }
  }
}
```

2. Record some data to that topic in an MCAP file.

3. Attempt to open the MCAP file in Foxglove Studio

**Expected Behavior**

File should open without issues and the topic's schema should be parsed correctly.

**Actual Behavior**
![Screenshot 2024-02-01 at 10 48 12 AM](https://github.com/foxglove/studio/assets/363800/3f9a0166-dc5a-4ecd-b6f3-69b4f442b147)

> Error in topic ... (channel 1): Unsupported type undefined for array item objects

**Comment by foxhubber** [comment]
Internal tracking ticket: [FG-6486](https://linear.app/foxglove/issue/FG-6486/foxglove-studio-does-not-appear-to-understand-json-schema-dollarref-in)

**Comment by defunctzombie** [comment]
Thanks for reporting. This is a known limitation and we have a better error message now.

**Reply by willat343** [reply]
Is there a plan to address this limitation in a future release? Or should we prefer other formats that support import (e.g. Protobuf) for creating custom schemas that are composed of "sub-schemas"?

**Reply by defunctzombie** [reply]
There's no plan to support $ref as described above right now and I definitely would encourage the use of formats like Protobuf for sub-schemas.
